diff --git a/0003-Fix-conftest-to-use-nv_drm_gem_vmap-which-has-the-se.patch b/0003-Fix-conftest-to-use-nv_drm_gem_vmap-which-has-the-se.patch
new file mode 100644
index 0000000..89f7038
--- /dev/null
+++ b/0003-Fix-conftest-to-use-nv_drm_gem_vmap-which-has-the-se.patch
@@ -0,0 +1,31 @@
+From: Benjamin ROBIN <dev@benjarobin.fr>
+Date: Sun, 12 May 2024 17:54:18 +0200
+Subject: [PATCH 3/3] Fix conftest to use nv_drm_gem_vmap() which has the
+ secondary map argument
+
+See https://forums.developer.nvidia.com/t/nvidia-modules-build-failure-with-upcoming-gcc-14-and-recent-kernels-due-to-misfiring-conftest-sh-test-heads-up/279072
+---
+ kernel/conftest.sh | 7 ++++++-
+ 1 file changed, 6 insertions(+), 1 deletion(-)
+
+diff --git a/kernel/conftest.sh b/kernel/conftest.sh
+index 678b79c..606f13d 100755
+--- a/kernel/conftest.sh
++++ b/kernel/conftest.sh
+@@ -4593,8 +4593,13 @@ compile_test() {
+             #
+             CODE="
+             #include <drm/drm_gem.h>
++            #if defined(NV_LINUX_IOSYS_MAP_H_PRESENT)
++            typedef struct iosys_map nv_sysio_map_t;
++            #else
++            typedef struct dma_buf_map nv_sysio_map_t;
++            #endif
+             int conftest_drm_gem_object_vmap_has_map_arg(
+-                    struct drm_gem_object *obj, struct dma_buf_map *map) {
++                    struct drm_gem_object *obj, nv_sysio_map_t *map) {
+                 return obj->funcs->vmap(obj, map);
+             }"
+ 
+-- 
+2.45.0
diff --git a/PKGBUILD b/PKGBUILD
index c68e6eb..c69a65d 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -1,27 +1,29 @@
 # Maintainer: archeese
 
 _pkgmainbranch=nvidia-utils
-pkgbase=nvidia-535xx-utils
+pkgbase=nvidia-535xx-grid
 pkgname=('nvidia-535xx-utils' 'opencl-nvidia-535xx' 'nvidia-535xx-dkms')
-pkgver=535.183.01
-pkgrel=2
+pkgver=535.161.07
+pkgrel=1
 pkgdesc="NVIDIA drivers for Linux, 535 branch, dkms"
 arch=('x86_64')
 url="http://www.nvidia.com/"
 license=('custom')
 options=('!strip')
 makedepends=('patchelf')
-_pkg="NVIDIA-Linux-x86_64-${pkgver}"
+_pkg="NVIDIA-Linux-x86_64-${pkgver}-grid"
 source=('nvidia-drm-outputclass.conf'
         'nvidia-utils.sysusers'
         'nvidia.rules'
         'linux-6.10.patch'
-        "https://us.download.nvidia.com/XFree86/Linux-x86_64/${pkgver}/${_pkg}.run")
+        '0003-Fix-conftest-to-use-nv_drm_gem_vmap-which-has-the-se.patch'
+        "https://storage.googleapis.com/nvidia-drivers-us-public/GRID/vGPU16.4/${_pkg}.run")
 sha512sums=('de7116c09f282a27920a1382df84aa86f559e537664bb30689605177ce37dc5067748acf9afd66a3269a6e323461356592fdfc624c86523bf105ff8fe47d3770'
             '4b3ad73f5076ba90fe0b3a2e712ac9cde76f469cd8070280f960c3ce7dc502d1927f525ae18d008075c8f08ea432f7be0a6c3a7a6b49c361126dcf42f97ec499'
             'a0ceb0a6c240cf97b21a2e46c5c212250d3ee24fecef16aca3dffb04b8350c445b9f4398274abccdb745dd0ba5132a17942c9508ce165d4f97f41ece02b0b989'
             '29b156ee3504f11216e462d69c219f1c144a36f19b6756301e3db133c4c471d0b94483fd2809a32909c43f9cbaadd4016988b55354e77204f7d8a0033e735a8e'
-            '02b6b679f4fc1d5305f32fca8ce0875eef04cb99f5611d0bb85ac7607ecdd5b2aa4d60b51bf47546477464531a07fffa5bf3db3859868648bd5e86565d85afbb')
+            'b17472e3b66a9018b49774cc45c71d401dd9454412fcd86bdae731f88f63a3c0aa463f4226a4b6349b1308225fe38471decb2bc5de4db141b1b16dede515398f'
+            'c7554a3b581ff70a5cfd3f2d4e370af9e18556defdaa6d318acc957d40283cd14969c923edd0cc2a95a54a4aaaf2ca67453ccd8026e28a3e0ed15fd1dd35fa17')
 
 
 create_links() {
@@ -42,6 +44,8 @@ prepare() {
     # https://aur.archlinux.org/packages/nvidia-535xx-dkms#comment-982769
     patch -Np1 -i "$srcdir/linux-6.10.patch"
 
+    patch -Np1 -i ../0003-Fix-conftest-to-use-nv_drm_gem_vmap-which-has-the-se.patch
+
     cd kernel
 
     sed -i "s/__VERSION_STRING/${pkgver}/" dkms.conf
@@ -235,11 +239,11 @@ package_nvidia-535xx-utils() {
     ln -s nvidia "${pkgdir}/usr/share/doc/nvidia-utils"
 
     # new power management support
-    install -Dm644 systemd/system/*.service -t "${pkgdir}/usr/lib/systemd/system"
-    install -Dm755 systemd/system-sleep/nvidia "${pkgdir}/usr/lib/systemd/system-sleep/nvidia"
-    install -Dm755 systemd/nvidia-sleep.sh "${pkgdir}/usr/bin/nvidia-sleep.sh"
-    install -Dm755 nvidia-powerd "${pkgdir}/usr/bin/nvidia-powerd"
-    install -Dm644 nvidia-dbus.conf "${pkgdir}"/usr/share/dbus-1/system.d/nvidia-dbus.conf
+#    install -Dm644 systemd/system/*.service -t "${pkgdir}/usr/lib/systemd/system"
+#    install -Dm755 systemd/system-sleep/nvidia "${pkgdir}/usr/lib/systemd/system-sleep/nvidia"
+#    install -Dm755 systemd/nvidia-sleep.sh "${pkgdir}/usr/bin/nvidia-sleep.sh"
+#    install -Dm755 nvidia-powerd "${pkgdir}/usr/bin/nvidia-powerd"
+#    install -Dm644 nvidia-dbus.conf "${pkgdir}"/usr/share/dbus-1/system.d/nvidia-dbus.conf
 
     # distro specific files must be installed in /usr/share/X11/xorg.conf.d
     install -Dm644 "${srcdir}/nvidia-drm-outputclass.conf" "${pkgdir}/usr/share/X11/xorg.conf.d/10-nvidia-drm-outputclass.conf"
